src/api/mod.rs

```diff
@@ -1,29 +1,160 @@
+        let lock = api.token.read().await;
+        let expire_in = lock.expire_in;
+        drop(lock);
+
+        // Refreshes in (3600 - 300) secs just in case
+        tokio::time::sleep(Duration::from_secs(expire_in - 300)).await;
+
+        let mut lock = api.token.write().await;
+
+        // Another loop to ensure token is actually updates
+        // (without any errors)
+        // If error anyway happened then it's going to retry
+        // until successed
+        loop {
+            let res = lock.refresh_token(&api.api_root, &api.client).await;
+
+            match res {
+                Ok(_) => break,
+                Err(e) => {
+                    println!("{:?}", e);
+                }
+            }
+        }
+
+        drop(lock)
+    }
+}
+
+pub async fn refresh_token_worker(api: Arc<OtrApiBody>, receiver: Receiver<()>) {
+    tokio::select! {
+        _ = refresh_token_loop(api) => {}
+        _ = receiver => {}
+    }
+}
+
+pub struct OtrToken {
+    pub token: String,
+    pub refresh_token: String,
+    pub expire_in: u64
+}
+
+impl OtrToken {
+    /// Function that refresh token when called
+    pub async fn refresh_token(&mut self, api_root: &str, client: &Client) -> Result<(), Error> {
+        let link = format!("{}/v1/oauth/refresh?refreshToken={}", api_root, self.refresh_token);
+
+        let mut response: OAuthResponse = client
+            .post(link)
+            .header(CONTENT_TYPE, "application/json")
+            .send()
+            .await?
+            .json()
+            .await?;
+
+        response.token.insert_str(0, "Bearer ");
+
+        self.token = response.token;
+        self.refresh_token = response.refresh_token;
+        self.expire_in = response.expire_in;
+
+        Ok(())
+    }
+}
+
+pub struct OtrApiBody {
     client: Client,
-    token: String,
-    api_root: String
+    api_root: String,
+
+    /// Wrapped in RwLock because we need
+    /// a shared mutable access
+    token: RwLock<OtrToken>
+}
+
+pub struct OtrApiClient {
+    /// Wrapped in [`Arc`] because everything located in [`OtrApiBody`]
+    /// need to be accessed in diffrent threads (shared reference)
+    body: Arc<OtrApiBody>,
+
+    /// Channel that's get sended on Drop to shutdown refresh token worker
+    /// Why it's wrapped in `[Option]`? Because oneshot sender
+    /// consumes itself on send.
+    ///
+    /// Detailed explanation:
+    /// In Drop trait implementation we having a mutable reference
+    /// on our struct (where sender is located), so when `send()`
+    /// happens it consumes itself, but we still having that mutable
+    /// reference and because of this we getting a compile error
+    /// variabled getting moved.
+    /// Workaround is pretty simple:
+    /// Wrap [`Sender`] inside Option, so we can use [`std::mem::take`]
+    /// to replace our sender with default value (in our case in None)
+    /// and peacefully let sender consume itself :)
```

Comment:
```suggestion
    /// Channel that gets sent on `Drop` to shut down the refresh token worker.
    /// Why is it wrapped in an `[Option]`? Because the oneshot sender
    /// consumes itself upon sending.
    ///
    /// Detailed explanation:
    /// In the `Drop` trait implementation, we have a mutable reference
    /// to our struct (where the sender is located). So, when `send()`
    /// occurs, it consumes itself, but we still hold that mutable
    /// reference, and because of this, we encounter a compile error
    /// indicating the variable has been moved.
    /// The workaround is pretty simple:
    ///	Wrap the [`Sender`] inside an `Option`, so we can use [`std::mem::take`]
    ///	to replace our sender with a default value (in our case, `None`)
    ///	and allow the sender to consume itself peacefully.
```